tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -472,4 +472,136 @@ TEST_F(
   }
 }
 
+class AGOverlapTest : public MultiDeviceTest {
+ protected:
+  OverlapTestParams params;
+
+  int64_t num_devices_;
+  int64_t my_device_index_;
+  std::vector<int64_t> all_devices_;
+  at::Tensor ta_, tb_, tc_;
+  at::Tensor ta_sharded_, tc_sharded_;
+  // stores the backend
+  c10d::Backend* world_communicator_;
+
+  // Define I/O and intermediate Tensor shapes
+  // A has shape (S, sharded(n), M/(S*n), K)
+  // B(K,N)
+  // C has shape (S, M/S, N)
+  std::vector<int64_t> ta_sizes;
+  std::vector<int64_t> tb_sizes;
+  std::vector<int64_t> tc_sizes;
+  std::vector<int64_t> ta_sharded_sizes;
+  std::vector<int64_t> tc_sharded_sizes;
+  std::vector<int64_t> tc_j_sizes;
+
+  void SetUp() {
+    MultiDeviceTest::SetUp();
+
+    num_devices_ = communicator_->size();
+    my_device_index_ = communicator_->deviceId();
+    ASSERT_EQ(params.M % num_devices_, 0);
+    ASSERT_EQ(params.S % num_devices_, 0);
+
+    // Setup the world communicators
+    std::vector<int64_t> devices(num_devices_);
+    std::iota(devices.begin(), devices.end(), 0);
+    all_devices_ = std::move(devices);
+    world_communicator_ =
+        communicator_->getBackendForTeam(all_devices_, params.backend_type);
+
+    // Debug print
+    if (communicator_->deviceId() == 0 && debug_print) {
+      debug() << params << std::endl;
+    }
+
+    ta_sizes = std::vector<int64_t>{
+        params.S, num_devices_, params.M / (num_devices_ * params.S), params.K};
```

Comment:
Can you point me to the source of this algorithm? What you try to implement is apparently different from [what I had in mind](https://docs.google.com/document/d/1BK_xEolbJDEky5trLW8jbuNHMX-v2QSY40O8782lYQ8/edit#heading=h.ao128ya3cf28). I don't have evidence showing which is better, but I suspect host-sharding the M dimension (in addition to device-sharding it already) may lead to wide submatrices that are less compute friendly. So I'm curious what people do in practice in e.g. TransformerEngine.

Reply:
There is no source, this is just what Sam and I put together from our meeting together: https://docs.google.com/document/d/1Fzr9Zs2Dqfj3e4yR8LKxFrRqC1EkMUfQczJMYQGJQUI/edit#heading=h.8niucpl6b50

Would it make sense for me to implement the algorithm you had in mind using ATen after this one is merged? I believe the main difference is that yours device shards along rows of A and host-shards along columns of B, where mine does both on A and fully replicates B.

Reply:
> I believe the main difference is that yours device shards along rows of A and host-shards along columns of B, where mine does both on A and fully replicates B.

There's that, but I don't worry about that difference too much. The main difference to me is that your current implementation host-shards A's rows and we could alternatively host-shard A's columns (while A remains to be device-sharded row-wise). 

Yes, a follow-up sounds good to me. It's certainly shape dependent.

Reply:
Tiling the reduced dimension is definitely possible, but it is not what is done in Transformer sequence parallel

Reply:
Let me try again. I was talking about host-sharding the reduced dimension, which I believe is what's done and/or what can be done for sequence parallel. For example:

Assume batch size is 1. 

```
input: [s_d, h]. the output of the second layernorm, which is sequence parallel
weight: [h, 4h_d]. the weight of the first linear layer in MLP, which is hidden dimension parallel
output: [s, 4h_d]

input' = allgather(input)
output = matmul(input', weight)
```

This can be host-sharded to S iterations. The first iteration is

```
output = 0
for i in range(S):
  input' = allgather(input[s_d, h/S*i:h/S*(i+1)])
  output' = matmul(input', weight[h/S*i:h/S*(i+1)])
  output += output'
```

In GPT-3, s=2048 and h=12288. So I think host-sharding h is preferable.

Reply:
I was not understanding correctly. You are not talking about sharding but about host-tiling those axis, nvm.

So, it is _possible_ to host-tile either the non-contracting or the contracting axis. It is important to support both eventually. The same holds for GEMM+RS. I suggest to merge this PR and to implement the other one later.

We started with tiling the non-contracting axis because **1) it is what is done in TE** and 2) it is simpler: when tiling the contracting dimension, we need to tile both the input and the weight, and to accumulate the result.

_Another practical point to keep in mind: the host-parallelilzed axis needs general to be at the outermost position in the allocation domain, otherwise the `allgather` is not possible as-is. The layout we chose is thus slightly more straightforward as it saves a transposition._

Thanks for the concrete dimension lengths, that's really helpful. I hear you: it is in general better to tile large dimension. However, I would be careful about what's best for e2e performance. For example, one element in favor of tiling the non-contracting axis: tiling the contracting axis requires a global synchronization at the end of the AG+GEMM, while the alternative potentially allows to push the synchronization point further, even possibly at the very end of the MLP, if I'm not mistaken.